File: c/yacht.c
#include "yacht.h"
#include <stdbool.h>
#include <stdlib.h>
#include <string.h>

static bool findCount(int counter[5], int ct) {
  for (int i = 0; i < 5; i++) {
    if (counter[i] == ct) {
      return true;
    }
  }
  return false;
}

static int cmp(const void *a, const void *b) {
  int int_a = *((int *)a);
  int int_b = *((int *)b);

  if (int_a == int_b)
    return 0;
  if (int_a < int_b)
    return -1;
  return 1;
}

int score(dice_t dice, category_t category) {
  qsort(dice.faces, 5, sizeof(int), cmp);

  int counter[6] = {0};
  int sum = 0;
  for (int i = 0; i < 5; i++) {
    counter[dice.faces[i] - 1]++;
    sum += dice.faces[i];
  }

  if (category == YACHT) {
    if (findCount(counter, 5)) {
      return 50;
    }
  }

  if (category < FULL_HOUSE) {
    return counter[category] * (category + 1);
  }

  if (category == FULL_HOUSE) {
    if (findCount(counter, 3) && findCount(counter, 2)) {
      return sum;
    }
  }

  if (category == FOUR_OF_A_KIND) {
    for (int i = 0; i < 6; i++) {
      if (counter[i] >= 4) {
        return (i + 1) * 4;
      }
    }
  }

  if (category == LITTLE_STRAIGHT) {
    int tmp[5] = {1, 2, 3, 4, 5};
    if (!memcmp(&dice.faces, &tmp, sizeof(int) * 5)) {
      return 30;
    }
  }

  if (category == BIG_STRAIGHT) {
    int tmp[5] = {2, 3, 4, 5, 6};
    if (!memcmp(&dice.faces, &tmp, sizeof(int) * 5)) {
      return 30;
    }
  }

  if (category == CHOICE) {
    return sum;
  }
  return 0;
}

File: dart/yacht.dart
import 'package:yacht/categories.dart';
import 'package:collection/collection.dart';

class Yacht {
  List<int> dice;
  final Map<int, int> counter = {};

  Yacht(this.dice) {
    for (final d in this.dice) {
      counter[d] = (counter[d] ?? 0) + 1;
    }
    dice.sort();
  }

  bool findCount(int n) {
    return counter.values.any((v) => v == n);
  }

  int score(Category category) {
    if (category.index <= Category.sixes.index) {
      final n = category.index + 1;
      return (counter[n] ?? 0) * n;
    }

    if (category == Category.yacht && findCount(5)) {
      return 50;
    }

    if (category == Category.full_house && findCount(3) && findCount(2)) {
      return dice.sum;
    }

    if (category == Category.four_of_a_kind) {
      for (final MapEntry(:key, :value) in counter.entries) {
        if (value >= 4) {
          return key * 4;
        }
      }
    }

    if (category == Category.little_straight &&
        ListEquality().equals(dice, [1, 2, 3, 4, 5])) {
      return 30;
    }

    if (category == Category.big_straight &&
        ListEquality().equals(dice, [2, 3, 4, 5, 6])) {
      return 30;
    }

    if (category == Category.choice) {
      return dice.sum;
    }
    return 0;
  }
}
